src/SignalR/clients/ts/signalr/src/HttpConnection.ts

```diff
@@ -349,19 +419,33 @@ export class HttpConnection implements IConnection {
         return transport && typeof (transport) === "object" && "connect" in transport;
     }
 
-    private changeState(from: ConnectionState, to: ConnectionState): boolean {
-        if (this.connectionState === from) {
-            this.connectionState = to;
-            return true;
-        }
-        return false;
-    }
-
     private stopConnection(error?: Error): void {
+        this.logger.log(LogLevel.Debug, `HttpConnection.stopConnection(${error}) called while in sate ${this.connectionState}.`);
+
+        const connectionStarted = this.connectionStarted;
+
         this.transport = undefined;
+        this.connectionStarted = false;
 
         // If we have a stopError, it takes precedence over the error from the transport
         error = this.stopError || error;
+        this.stopError = undefined;
+
+        if (this.connectionState === ConnectionState.Disconnected) {
+            this.logger.log(LogLevel.Debug, `Call to HttpConnection.stopConnection(${error}) was ignored because the connection is already in the disconnected state.`);
+            return;
+        }
+
+        if (this.connectionState === ConnectionState.Connecting) {
+            this.logger.log(LogLevel.Warning, `Call to HttpConnection.stopConnection(${error}) was ignored because the connection hasn't yet left the in the connecting state.`);
```

Comment:
I don't expect this will ever be logged since we just finished awaiting startInternal() here. startInternal() either:

1. Completes after successfully transitioning from the Connecting to Connected state.
2. Fails after moving back into the Disconnected state.
3. Sees that stop/stopInternal() transitioned the connection to the Disconnecting state and exits while still in the Disconnecting state. This is with the realization that startInternal() needs to complete for stopInternal to continue and the expectation that start() will await the stopInternal() promise and only throw after the connection has fully transition into the Disconnected state.

This check and log might be overly defensive, but the logic is complicated enough I'd like to be able to know this would be logged if and when we investigate any connection state issues.